File: figures/agent/render_F10_agent_envelope.py
#!/usr/bin/env python3
"""Render F10 - main-text Agent figure (three panels).

Scientific design frozen in docs/F10_AGENT_FIGURE_SPEC_AND_ED_SI_PLAN_2026-09-13.md.
Metric definitions fixed by docs/AGENT_METRIC_DEFINITIONS_SOURCE_OF_TRUTH_2026-09-13.md.

This script performs RENDERING ONLY. Every plotted value is read from
data/agent_figure_panel_data_2026-09-13.csv (built by tools/discover/build_agent_figure_data.py) and from
data/discover_boundary_c1_error_taxonomy_summary.csv for the median smallest-window series. Nothing is hand-entered.

Usage (from the repository root):
  python figures/agent/render_F10_agent_envelope.py
"""
from __future__ import annotations
import csv
import hashlib
from pathlib import Path

import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.lines import Line2D

ROOT = Path(__file__).resolve().parents[2]
PANEL = ROOT / "data" / "agent_figure_panel_data_2026-09-13.csv"
TAX = ROOT / "data" / "discover_boundary_c1_error_taxonomy_summary.csv"
OUT = ROOT / "figures" / "agent"
STEM = "F10_agent_capability_bounded_envelope"

D_THRESHOLD = 206.0          # policy D full-decision completion threshold, CU
LOWEST_STABLE = 75.0         # lowest strong budget with 20/20 completion, CU
FULL_DOMAIN = 14136          # admissible process states
NON_BINDING = 5000.0         # the non-binding allowance, plotted on its own axis

COL = {"strong": "#1f77b4", "mini": "#ff7f0e", "nano": "#d62728"}
INK = "#1a1a1a"
GREY = "#5a5a5a"
BAND = "#9ecae1"


def load(path: Path) -> list[dict]:
    with path.open(encoding="utf-8") as f:
        return list(csv.DictReader(f))


def fnum(v):
    return None if v in ("", None) else float(v)


def main() -> int:
    rows = load(PANEL)
    tax = {(r["tier"], r["arm"], int(float(r["budget_CU"]))): r for r in load(TAX)}

    def cells(tier, arm="E", exclude_non_binding=True):
        out = [r for r in rows if r["tier"] == tier and r["arm"] == arm]
        if exclude_non_binding:
            out = [r for r in out if not int(r["is_non_binding_allowance"])]
        return sorted(out, key=lambda r: int(r["budget_CU"]))

    nb = next(r for r in rows if int(r["is_non_binding_allowance"]))

    plt.rcParams.update({
        "font.size": 9, "axes.labelsize": 9, "axes.titlesize": 9,
        "xtick.labelsize": 8, "ytick.labelsize": 8, "legend.fontsize": 7.4,
        "axes.edgecolor": INK, "axes.labelcolor": INK, "text.color": INK,
        "xtick.color": INK, "ytick.color": INK, "axes.linewidth": 0.8,
        "savefig.bbox": "tight", "figure.dpi": 150,
    })

    fig = plt.figure(figsize=(13.6, 4.0))
    gs = fig.add_gridspec(1, 3, wspace=0.50, left=0.050, right=0.988, bottom=0.16, top=0.88)
    axes = []
    for i in range(3):
        sub = gs[0, i].subgridspec(1, 2, width_ratios=[5.2, 1.0], wspace=0.06)
        axes.append((fig.add_subplot(sub[0, 0]), fig.add_subplot(sub[0, 1])))

    def style_pair(ax, axn, ylabel, ylim=None, logy=False):
        for a in (ax, axn):
            a.grid(True, which="major", lw=0.4, color="#d9d9d9", zorder=0)
            a.set_axisbelow(True)
            a.tick_params(direction="out", length=3)
        ax.set_xscale("log")
        ax.set_xlim(42, 470)
        ax.set_xticks([50, 75, 100, 150, 200, 250, 400])
        ax.set_xticklabels(["50", "75", "100", "150", "200", "250", "400"])
        ax.minorticks_off()
        axn.set_xlim(0, 1)
        axn.set_xticks([0.5])
        axn.set_xticklabels(["non-binding\n5000 CU"], fontsize=7.2)
        axn.tick_params(axis="y", labelleft=False, length=0)
        axn.spines["left"].set_linestyle((0, (2, 2)))
        if logy:
            ax.set_yscale("log")
            axn.set_yscale("log")
        if ylim:
            ax.set_ylim(*ylim)
            axn.set_ylim(*ylim)
        ax.set_ylabel(ylabel)
        ax.set_xlabel("compute budget (CU)")

    def vrefs(ax, labels=True):
        ax.axvline(D_THRESHOLD, color=GREY, lw=0.9, ls="--", zorder=1)
        ax.axvline(LOWEST_STABLE, color=COL["strong"], lw=0.9, ls=":", zorder=1)
        if labels:
            ax.annotate("policy D threshold 206 CU", xy=(D_THRESHOLD, 0.985),
                        xycoords=("data", "axes fraction"), xytext=(-3, 0), textcoords="offset points",
                        fontsize=6.8, color=GREY, ha="right", va="top", rotation=90)
            ax.annotate("lowest stable strong budget 75 CU", xy=(LOWEST_STABLE, 0.015),
                        xycoords=("data", "axes fraction"), xytext=(4, 0), textcoords="offset points",
                        fontsize=6.8, color=COL["strong"], ha="left", va="bottom", rotation=90)

    # ---------------------------------------------------------------- panel a
    ax, axn = axes[0]
    style_pair(ax, axn, "P(complete decision)", ylim=(-0.06, 1.09))
    vrefs(ax)
    for tier, marker in (("strong", "o"), ("mini", "s"), ("nano", "^")):
        cs = cells(tier)
        x = [float(r["budget_CU"]) for r in cs]
        y = [float(r["p_complete_decision"]) for r in cs]
        lo = [y[i] - float(cs[i]["ci_lo"]) for i in range(len(cs))]
        hi = [float(cs[i]["ci_hi"]) - y[i] for i in range(len(cs))]
        ax.errorbar(x, y, yerr=[lo, hi], color=COL[tier], marker=marker, ms=4.2, lw=1.3,
                    capsize=2.2, elinewidth=0.8, zorder=3, label=f"{tier} (policy E)")
    y = float(nb["p_complete_decision"])
    axn.errorbar([0.5], [y], yerr=[[y - float(nb["ci_lo"])], [float(nb["ci_hi"]) - y]],
                 color=COL["strong"], marker="o", ms=4.2, capsize=2.2, elinewidth=0.8, zorder=3)
    floor = next(r for r in cells("strong") if int(r["budget_CU"]) == 50)
    ax.annotate("affordability floor at 50 CU\n13/20 complete; winner and\ndecision pair still 20/20",
                xy=(50, float(floor["p_complete_decision"])), xycoords="data",
                xytext=(0.40, 0.235), textcoords="axes fraction",
                fontsize=6.8, color=INK, ha="left", va="bottom",
                arrowprops=dict(arrowstyle="-", lw=0.6, color=GREY,
                                connectionstyle="angle,angleA=0,angleB=90,rad=3"))
    ax.set_title("a   decision recovery and capability boundary", loc="left", fontweight="bold")

    # ---------------------------------------------------------------- panel b
    ax, axn = axes[1]
    style_pair(ax, axn, "fraction of runs using\nnarrow-window allocation", ylim=(-0.06, 1.09))
    vrefs(ax, labels=False)
    for tier, marker in (("strong", "o"), ("mini", "s"), ("nano", "^")):
        cs = [r for r in cells(tier) if r["narrow_window_fraction"] not in ("", None)]
        ax.plot([float(r["budget_CU"]) for r in cs], [float(r["narrow_window_fraction"]) for r in cs],
                color=COL[tier], marker=marker, ms=4.2, lw=1.3, zorder=3, label=f"{tier} (policy E)")
    axn.plot([0.5], [float(nb["narrow_window_fraction"])], color=COL["strong"], marker="o", ms=4.2, zorder=3)
    axw = ax.twinx()
    axw.set_yscale("log")
    axw.set_ylim(380, 40000)
    axw.set_ylabel(f"median smallest window\n(states, of {FULL_DOMAIN:,})", color=GREY)
    axw.tick_params(axis="y", colors=GREY, labelsize=8, direction="out", length=3)
    axw.spines["right"].set_color(GREY)
    wx, wy = [], []
    for r in cells("strong"):
        t = tax.get(("strong", "E", int(float(r["budget_CU"]))))
        v = fnum(t["min_window_states_median"]) if t else None
        if v is not None:
            wx.append(float(r["budget_CU"]))
            wy.append(v)
    axw.scatter(wx, wy, s=16, facecolor="none", edgecolor=GREY, lw=0.9, zorder=2, marker="D")
    axw.axhline(FULL_DOMAIN, color=GREY, lw=0.6, ls="-.", zorder=1)
    axw.annotate(f"full domain {FULL_DOMAIN:,} (no narrowing)", xy=(0.97, FULL_DOMAIN),
                 xycoords=("axes fraction", "data"), xytext=(0, 4), textcoords="offset points",
                 fontsize=6.8, color=GREY, ha="right")
    ax.set_title("b   allocation mechanism", loc="left", fontweight="bold")
    ax.legend(loc="center left", frameon=False, handlelength=1.6, bbox_to_anchor=(0.015, 0.60))
    axw.legend(handles=[Line2D([], [], marker="D", color=GREY, lw=0, mfc="none", ms=4)],
               labels=["median smallest window (right axis)"], loc="center left", frameon=False,
               bbox_to_anchor=(0.015, 0.40), handlelength=1.2, fontsize=7.0, labelcolor=GREY)

    # ---------------------------------------------------------------- panel c
    ax, axn = axes[2]
    style_pair(ax, axn, "CU spent", ylim=(30, 9000), logy=True)
    ax.axvline(D_THRESHOLD, color=GREY, lw=0.9, ls="--", zorder=1)
    ax.axvline(LOWEST_STABLE, color=COL["strong"], lw=0.9, ls=":", zorder=1)
    for a in (ax, axn):
        a.axhline(D_THRESHOLD, color=GREY, lw=0.9, ls="--", zorder=1)
    cs = [r for r in cells("strong") if r["decision_stable_CU_median"] not in ("", None)]
    xs = [float(r["budget_CU"]) for r in cs]
    st = [float(r["decision_stable_CU_median"]) for r in cs]
    fu = [float(r["final_used_CU_median"]) for r in cs]
    for x, s_, f_ in zip(xs, st, fu):
        ax.plot([x, x], [s_, f_], color=BAND, lw=3.2, solid_capstyle="butt", zorder=2)
    ax.plot(xs, st, color=COL["strong"], marker="o", ms=4.2, lw=1.3, zorder=3, label="median decision-stable CU")
    ax.plot(xs, fu, color=INK, marker="v", ms=4.2, lw=1.3, zorder=3, label="median final-used CU")
    s_nb, f_nb = float(nb["decision_stable_CU_median"]), float(nb["final_used_CU_median"])
    axn.plot([0.5, 0.5], [s_nb, f_nb], color=BAND, lw=3.2, solid_capstyle="butt", zorder=2)
    axn.plot([0.5], [s_nb], color=COL["strong"], marker="o", ms=4.2, zorder=3)
    axn.plot([0.5], [f_nb], color=INK, marker="v", ms=4.2, zorder=3)
    nb_max = float(nb["final_used_CU_range"].split("-")[1])
    axn.plot([0.5], [nb_max], color=INK, marker="v", ms=4.6, mfc="none", mew=0.9, zorder=3)
    axn.annotate(f"{nb_max:.0f} CU\n(single run)", xy=(0.5, nb_max), xytext=(0, 7),
                 textcoords="offset points", fontsize=6.6, color=INK, ha="center", va="bottom")
    ax.annotate(f"non-binding allowance: median {f_nb:.0f} CU = {f_nb / D_THRESHOLD:.1f}x D",
                xy=(0.985, 0.96), xycoords="axes fraction", fontsize=6.9, color=INK, ha="right", va="top")
    ax.annotate("206 CU", xy=(0.02, D_THRESHOLD), xycoords=("axes fraction", "data"),
                xytext=(0, 3), textcoords="offset points", fontsize=6.8, color=GREY)
    ax.set_title("c   upper cost boundary", loc="left", fontweight="bold")
    ax.legend(loc="upper left", frameon=False, handlelength=1.6, bbox_to_anchor=(0.015, 0.88))

    OUT.mkdir(parents=True, exist_ok=True)
    written = []
    for ext, kw in (("svg", {}), ("pdf", {}), ("png", {"dpi": 600})):
        f = OUT / f"{STEM}.{ext}"
        fig.savefig(f, **kw)
        written.append(f)
    plt.close(fig)

    man = OUT / "F10_RENDER_SHA256.txt"
    lines = [f"# F10 render manifest - {STEM}",
             "# spec: docs/F10_AGENT_FIGURE_SPEC_AND_ED_SI_PLAN_2026-09-13.md", ""]
    for f in written + [PANEL, TAX, Path(__file__)]:
        h = hashlib.sha256(f.read_bytes()).hexdigest()
        lines.append(f"{h}  {f.relative_to(ROOT).as_posix()}")
    man.write_text("\n".join(lines) + "\n", encoding="utf-8")

    for f in written:
        print(f"wrote {f.relative_to(ROOT).as_posix()}  ({f.stat().st_size:,} bytes)")
    print(f"wrote {man.relative_to(ROOT).as_posix()}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: figures/agent/render_F10_agent_envelope.R
#!/usr/bin/env Rscript

# Figure 10 — Agent capability-bounded operating envelope with oracle normalization
# 2026-09-20 supervisor update. Reads frozen C1 panel data + oracle summary only.

args <- commandArgs(trailingOnly = TRUE)
root <- if (length(args) >= 1) args[[1]] else "."
panel_file <- file.path(root, "data", "agent_figure_panel_data_2026-09-13.csv")
window_file <- file.path(root, "analysis", "supervisor_2026_09_20", "agent_window_summary.csv")
oracle_file <- file.path(root, "analysis", "supervisor_2026_09_20", "agent_oracle_summary.csv")
out_dir <- file.path(root, "figures", "agent")
dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)

p <- read.csv(panel_file, stringsAsFactors = FALSE, check.names = FALSE)
w <- read.csv(window_file, stringsAsFactors = FALSE)
o <- read.csv(oracle_file, stringsAsFactors = FALSE)
ov <- setNames(o$value, o$metric)

ORACLE <- ov[["protocol_oracle_CU"]]; D <- ov[["D_threshold_CU"]]
if (abs(ORACLE - 22) > 1e-12) stop("Protocol oracle anchor changed")
if (abs(D - 206) > 1e-12) stop("Policy-D threshold changed")

ink <- "#111111"; blue <- "#315A8A"; orange <- "#B45F3B"; red <- "#8F3B3B"; green <- "#447A6A"; grey <- "#666666"; light <- "#E6E6E6"
strong <- p[p$tier == "strong" & p$arm == "E" & p$is_non_binding_allowance == 0, ]
strong <- strong[order(strong$budget_CU), ]
mini <- p[p$tier == "mini" & p$arm == "E", ]; mini <- mini[order(mini$budget_CU), ]
nano <- p[p$tier == "nano" & p$arm == "E", ]; nano <- nano[order(nano$budget_CU), ]
nb <- p[p$tier == "strong" & p$is_non_binding_allowance == 1, ][1, ]

render <- function(device = c("svg", "pdf", "png")) {
  device <- match.arg(device)
  stem <- file.path(out_dir, "F10_agent_capability_bounded_envelope")
  if (device == "svg") svg(paste0(stem, ".svg"), width = 12.2, height = 4.2, pointsize = 9, bg = "white")
  if (device == "pdf") pdf(paste0(stem, ".pdf"), width = 12.2, height = 4.2, pointsize = 9, useDingbats = FALSE)
  if (device == "png") png(paste0(stem, ".png"), width = 12.2, height = 4.2, units = "in", res = 500, pointsize = 9, bg = "white")
  on.exit(dev.off(), add = TRUE)
  old <- par(no.readonly = TRUE); on.exit(par(old), add = TRUE)
  layout(matrix(1:4, nrow = 1), widths = c(1.35, 1.35, 1.35, 0.40))
  par(family = "serif", fg = ink, col.axis = ink, col.lab = ink, col.main = ink,
      mgp = c(2.2, 0.7, 0), tcl = -0.25, las = 1)

  # a — completion envelope
  par(mar = c(4.1, 4.3, 2.5, 0.8))
  plot(strong$budget_CU, strong$p_complete_decision, type = "o", log = "x", xlim = c(45, 430), ylim = c(-0.03, 1.06),
       xlab = "Compute budget (CU)", ylab = "P(complete decision)", col = blue, pch = 16, lwd = 1.4, axes = FALSE)
  abline(h = seq(0, 1, 0.2), col = light, lwd = 0.7)
  lines(mini$budget_CU, mini$p_complete_decision, type = "o", col = orange, pch = 15, lwd = 1.2)
  lines(nano$budget_CU, nano$p_complete_decision, type = "o", col = red, pch = 17, lwd = 1.2)
  abline(v = 75, col = blue, lty = 3, lwd = 1.0); abline(v = D, col = grey, lty = 2, lwd = 1.0)
  axis(1, at = c(50, 75, 100, 150, 200, 250, 400), labels = c("50", "75", "100", "150", "200", "250", "400")); axis(2); box(bty = "l")
  legend("bottomright", legend = c("strong", "mini", "nano"), col = c(blue, orange, red), pch = c(16, 15, 17), lty = 1, bty = "n", cex = 0.78)
  text(52, 0.96, sprintf("75 CU = %.2fx oracle\n206 CU = %.2fx oracle", 75 / ORACLE, D / ORACLE), adj = c(0, 1), cex = 0.72)
  mtext("a   Capability-bounded decision recovery", side = 3, adj = 0, line = 0.8, font = 2, cex = 0.95)

  # b — narrow-window mechanism and window size
  par(mar = c(4.1, 4.3, 2.5, 4.5))
  plot(strong$budget_CU, strong$narrow_window_fraction, type = "o", log = "x", xlim = c(45, 430), ylim = c(-0.03, 1.06),
       xlab = "Compute budget (CU)", ylab = "Fraction using narrow window", col = blue, pch = 16, lwd = 1.4, axes = FALSE)
  abline(h = seq(0, 1, 0.2), col = light, lwd = 0.7); abline(v = D, col = grey, lty = 2, lwd = 1.0)
  axis(1, at = c(50, 75, 100, 150, 200, 250, 400), labels = c("50", "75", "100", "150", "200", "250", "400")); axis(2); box(bty = "l")
  par(new = TRUE)
  plot(w$budget_CU[w$budget_CU < 5000], w$median_smallest_window_states[w$budget_CU < 5000], log = "xy",
       xlim = c(45, 430), ylim = c(500, 30000), type = "p", pch = 23, bg = "white", col = grey,
       axes = FALSE, xlab = "", ylab = "")
  axis(4, col.axis = grey); mtext("Median smallest window (states)", side = 4, line = 2.8, col = grey)
  abline(h = 14136, col = grey, lty = 3, lwd = 0.8)
  text(53, 0.05, "non-binding: narrow-window 0/20", adj = c(0, 0), cex = 0.72)
  mtext("b   Binding budgets activate scoped process search", side = 3, adj = 0, line = 0.8, font = 2, cex = 0.95)

  # c — stable/final spend + oracle
  par(mar = c(4.1, 4.3, 2.5, 0.8))
  ok <- !is.na(strong$decision_stable_CU_median)
  ss <- strong[ok, ]
  plot(ss$budget_CU, ss$decision_stable_CU_median, type = "o", log = "xy", xlim = c(45, 430), ylim = c(18, 950),
       xlab = "Compute budget (CU)", ylab = "CU spent", col = blue, pch = 16, lwd = 1.4, axes = FALSE)
  abline(h = pretty(c(20, 900)), col = light, lwd = 0.7)
  for (i in seq_len(nrow(ss))) segments(ss$budget_CU[i], ss$decision_stable_CU_median[i], ss$budget_CU[i], ss$final_used_CU_median[i], col = "#DCE6EF", lwd = 4)
  lines(ss$budget_CU, ss$decision_stable_CU_median, type = "o", col = blue, pch = 16, lwd = 1.4)
  lines(ss$budget_CU, ss$final_used_CU_median, type = "o", col = ink, pch = 25, bg = ink, lwd = 1.2)
  abline(h = ORACLE, col = green, lwd = 1.2); abline(h = D, col = grey, lty = 2, lwd = 1.0)
  axis(1, at = c(50, 75, 100, 150, 200, 250, 400), labels = c("50", "75", "100", "150", "200", "250", "400")); axis(2); box(bty = "l")
  legend("topleft", legend = c("decision-stable", "final used", "22-CU protocol oracle"), col = c(blue, ink, green), pch = c(16, 25, NA), lty = 1, pt.bg = c(NA, ink, NA), bty = "n", cex = 0.72)
  text(405, 23, sprintf("75-cell stable 52.5 CU = %.2fx\nD threshold 206 CU = %.2fx\nnon-binding stable 566 CU = %.2fx\n+148 CU -> 714 CU final", 52.5 / ORACLE, D / ORACLE, 566 / ORACLE), adj = c(1, 0), cex = 0.70)
  mtext("c   Oracle-normalized lower and upper compute boundaries", side = 3, adj = 0, line = 0.8, font = 2, cex = 0.95)

  # non-binding control, deliberately separate from the budget continuum
  par(mar = c(4.1, 1.0, 2.5, 1.0))
  plot(c(1, 1), c(nb$decision_stable_CU_median, nb$final_used_CU_median), type = "n", log = "y", xlim = c(0.7, 1.3), ylim = c(18, 5000), axes = FALSE, xlab = "", ylab = "")
  segments(1, nb$decision_stable_CU_median, 1, nb$final_used_CU_median, col = "#DCE6EF", lwd = 6)
  points(1, nb$decision_stable_CU_median, pch = 16, col = blue); points(1, nb$final_used_CU_median, pch = 25, bg = ink, col = ink)
  points(1, ov[["nonbinding_max_single_run_CU"]], pch = 25, bg = "white", col = ink)
  abline(h = ORACLE, col = green, lwd = 1.0); abline(h = D, col = grey, lty = 2, lwd = 0.9)
  axis(1, at = 1, labels = "non-binding\n5000 CU", tick = FALSE, cex.axis = 0.72)
  box(bty = "l", lty = 3)
  text(1, ov[["nonbinding_max_single_run_CU"]] * 1.10, "3021\n(single run)", cex = 0.62)
}

render("svg"); render("pdf"); render("png")
cat("Rendered F10 with 22-CU protocol oracle and non-binding 566 + 148 decomposition.\n")
